src/Http/ReceivedAuthnRequestQueryString.php

```diff
@@ -0,0 +1,292 @@
+<?php
+
+/**
+ * Copyright 2017 SURFnet B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Surfnet\SamlBundle\Http;
+
+use Surfnet\SamlBundle\Exception\RuntimeException;
+use Surfnet\SamlBundle\Http\Exception\InvalidReceivedAuthnRequestQueryStringException;
+use Surfnet\SamlBundle\Http\Exception\InvalidRequestException;
+
+final class ReceivedAuthnRequestQueryString
+{
+    const PARAMETER_REQUEST = 'SAMLRequest';
+    const PARAMETER_SIGNATURE = 'Signature';
+    const PARAMETER_SIGNATURE_ALGORITHM = 'SigAlg';
+    const PARAMETER_RELAY_STATE = 'RelayState';
+
+    private static $samlParameters = [
+        self::PARAMETER_REQUEST,
+        self::PARAMETER_SIGNATURE,
+        self::PARAMETER_SIGNATURE_ALGORITHM,
+        self::PARAMETER_RELAY_STATE,
+    ];
+
+    /**
+     * @var string
+     */
+    private $samlRequest;
+
+    /**
+     * @var string|null
+     */
+    private $signature;
+
+    /**
+     * @var string|null
+     */
+    private $signatureAlgorithm;
+
+    /**
+     * @var string|null
+     */
+    private $relayState;
+
+    private function __construct($samlRequest)
+    {
+        $this->samlRequest = $samlRequest;
+    }
+
+    /**
+     * @param string $requestUri
+     * @return ReceivedAuthnRequestQueryString
+     *
+     * @SuppressWarnings(PHPMD.CyclomaticComplexity) Extensive validation
+     * @SuppressWarnings(PHPMD.NPathComplexity) Extensive validation
+     */
+    public static function parse($requestUri)
+    {
+        if (!is_string($requestUri) || strlen($requestUri) <= strlen('?SAMLRequest=')) {
+            throw new InvalidReceivedAuthnRequestQueryStringException(sprintf(
+                'Could not parse query string: expected a non-empty string of at least %d characters, %s given',
+                strlen('?SAMLRequest='),
+                is_object($requestUri) ? get_class($requestUri) : gettype($requestUri)
+            ));
+        }
+
+        if (strpos($requestUri, '?') === false) {
```

Comment:
if it's a querystring, it should be the first character, or optional first character which is removed with an `ltrim`. Makes validation easier.